test/correctness/atomic_tuples.cpp

```diff
@@ -0,0 +1,224 @@
+        count_atomics++;
+        if (!op->mutex_name.empty()) {
+            count_atomics_with_mutexes++;
+        }
+        return IRMutator::visit(op);
+    }
+
+public:
+    int count_atomics = 0;
+    int count_atomics_with_mutexes = 0;
+};
+
+int main(int argc, char **argv) {
+    {
+        Func f, g;
+        Var x, y;
+
+        f(x, y) = {x, y};
+        f(x, y) = {f(x, y)[0] + x,
+                   f(x, y)[1] + y};
+
+        // The summation is independent in the two tuple components,
+        // so it can just be two atomic add instructions. No CAS loop
+        // required.
+        f.compute_root().update().parallel(y).atomic();
+
+        g(x, y) = f(x, y)[0] + f(x, y)[1];
+
+        Checker checker;
+        g.add_custom_lowering_pass(&checker, []() {});
+
+        Buffer<int> out = g.realize(128, 128);
+        for (int y = 0; y < 128; y++) {
+            for (int x = 0; x < 128; x++) {
+                int correct = 2 * x + 2 * y;
+                if (out(x, y) != correct) {
+                    printf("out(%d, %d) = %d instead of %d\n",
+                           x, y, out(x, y), correct);
+                    return -1;
+                }
+            }
+        }
+
+        if (checker.count_atomics != 2 || checker.count_atomics_with_mutexes != 0) {
+            printf("Expected two atomic nodes, neither of them with mutexes\n");
+            return -1;
+        }
+    }
+
+    {
+        Func f, g;
+        Var x, y;
+
+        f(x, y) = {x, y};
+        f(x, y) = {f(x, y)[1] + x,
+                   f(x, y)[0] + y};
+
+        // The summation is coupled across the two tuple components
+        // and there are two stores, so we need a mutex.
+        f.compute_root().update().parallel(y).atomic();
+
+        g(x, y) = f(x, y)[0] + f(x, y)[1];
+
+        Checker checker;
+        g.add_custom_lowering_pass(&checker, []() {});
+
+        Buffer<int> out = g.realize(128, 128);
+        for (int y = 0; y < 128; y++) {
+            for (int x = 0; x < 128; x++) {
+                int correct = 2 * x + 2 * y;
+                if (out(x, y) != correct) {
+                    printf("out(%d, %d) = %d instead of %d\n",
+                           x, y, out(x, y), correct);
+                    return -1;
+                }
+            }
+        }
+
+        if (checker.count_atomics != 1 || checker.count_atomics_with_mutexes != 1) {
+            printf("Expected one atomic node, with mutex\n");
+            return -1;
+        }
+    }
+
+    {
+        Func f, g;
+        Var x, y;
+
+        f(x, y) = {x, y, 0};
+        f(x, y) = {f(x, y)[1] + x,
+                   f(x, y)[0] + y,
+                   f(x, y)[2] + 1};
+
+        // The summation is coupled across the first two tuple
+        // components and there are two stores, so we need a mutex
+        // there. The last store could in principle be a separate atomic
+        // add, but we instead just pack it into the critical section.
```

Comment:
is this a missing optimization opportunity that should be documented as a github issue?

Reply:
I'm not sure that it's actually an optimization

Reply:
ah yes i agree